tools/mascot/src/main/java/dk/alexandra/fresco/tools/mascot/field/FieldElementUtils.java

```diff
@@ -0,0 +1,192 @@
+    BigInteger two = new BigInteger("2");
+    BigInteger current = BigInteger.ONE;
+    for (int i = 0; i < modBitLength; i++) {
+      generators.add(new FieldElement(current, modulus, modBitLength));
+      current = current.multiply(two).mod(modulus);
+    }
+    return generators;
+  }
+
+  /**
+   * Multiplies two lists of field elements, pair-wise.
+   * 
+   * @param leftFactors left factors
+   * @param rightFactors right factors
+   * @return list of products
+   */
+  public List<FieldElement> pairWiseMultiply(List<FieldElement> leftFactors,
+      List<FieldElement> rightFactors) {
+    if (leftFactors.size() != rightFactors.size()) {
+      throw new IllegalArgumentException("Lists must be same size");
+    }
+    return pairWiseMultiplyStream(leftFactors, rightFactors).collect(Collectors.toList());
+  }
+
+  /**
+   * Multiplies each value in list by scalar.
+   * 
+   * @param values list of factors
+   * @param scalar scalar factor
+   * @return list of products
+   */
+  public List<FieldElement> scalarMultiply(List<FieldElement> values, FieldElement scalar) {
+    return values.stream().map(value -> value.multiply(scalar)).collect(Collectors.toList());
+  }
+
+  /**
+   * Multiplies two lists of field elements, pair-wise.
+   * 
+   * @param leftFactors left factors
+   * @param rightFactors right factors
+   * @return stream of products
+   */
+  public Stream<FieldElement> pairWiseMultiplyStream(List<FieldElement> leftFactors,
+      List<FieldElement> rightFactors) {
+    return IntStream.range(0, leftFactors.size()).mapToObj(idx -> {
+      FieldElement l = leftFactors.get(idx);
+      FieldElement r = rightFactors.get(idx);
+      return l.multiply(r);
+    });
+  }
+
+  /**
+   * Computes inner product of two lists of field elements.
+   * 
+   * @param left left factors
+   * @param right right factors
+   * @return inner product
+   */
+  public FieldElement innerProduct(List<FieldElement> left, List<FieldElement> right) {
+    if (left.size() != right.size()) {
+      throw new IllegalArgumentException("Lists must have same size");
+    }
+    return sum(pairWiseMultiplyStream(left, right));
+  }
+
+  /**
+   * Computes inner product of elements and powers of twos.<br>
+   * e0 * 2**0 + e1 * 2**1 + ... + e(n - 1) * 2**(n - 1)
+   * Elements must have same modulus, otherwise we get undefined behaviour.
+   * 
+   * @param elements elements to recombine
+   * @return recombined elements
+   */
+  public FieldElement recombine(List<FieldElement> elements) {
+    if (elements.size() > modBitLength) {
+      throw new IllegalArgumentException("Number of elements cannot exceed bit-length");
+    }
+    BigInteger elementModulus = elements.get(0).getModulus();
+    if (!elementModulus.equals(modulus)) {
+      throw new IllegalArgumentException("Wrong modulus " + elementModulus);
+    }
+    return innerProduct(elements, generators.subList(0, elements.size()));
```

Comment:
Multiplying by _2<sup>k</sub>_ is really just bitshifting by _k_ places. That may be more efficient here.

Reply:
Currently not a bottle neck, but definitely a nice future optimization.